### ISSUE TITLE
Credentials class initializer should use make_creds instead of using the signet base class

### ISSUE THREAD

**Issue by 10io** [googleapis/google-auth-library-ruby#466]
Hello! 👋 

## 🔥 Problem

I've been trying to use the [Artifact Registry ruby library](https://github.com/googleapis/google-cloud-ruby/blob/139628702ad1c49be0fe7a3f5b504040ff680cdb/google-cloud-artifact_registry-v1/lib/google/cloud/artifact_registry/v1/artifact_registry/client.rb) with a [Workload Identity Federation](https://cloud.google.com/iam/docs/workload-identity-federation) authentication.

On top of this, I have to contact a server (metadata server?) to get the subject token (type JWT).

I was happy to see that I could build a credentials json structure with all these requirements:

```ruby
credentials = {
  'type' => 'external_account',
  'audience' => '<identity pool url>',
  'token_url' => 'https://sts.googleapis.com/v1/token',
  'subject_token_type' => 'urn:ietf:params:oauth:token-type:jwt',
  'credential_source' => {
    'url' => '<metadata server url>',
    'headers' => { 'Authorization' => "bearer <whatever is required to authenticate on the metadata server>" },
    'format' => { 'type' => 'json', 'subject_token_field_name' => 'token' } # <- this is awesome by the way
  }
}
```

As you can see, I'm programmatically building the credentials json structure which well, is a simple hash.

Now, I wanted to use this to build the Artifact Registry client:

```ruby
client = ::Google::Cloud::ArtifactRegistry::V1::ArtifactRegistry::Client.new do |config|
  config.credentials = credentials
end
```

Well, this doesn't work:

```
TypeError: no implicit conversion of nil into String
from /Users/me/.asdf/installs/ruby/3.1.4/lib/ruby/gems/3.1.0/gems/openssl-3.1.0/lib/openssl/pkey.rb:356:in `initialize'
```

I think the main problem here is that the [`Google::Auth::Credentials` initialize](https://github.com/googleapis/google-auth-library-ruby/blob/2a86821e1b2afc061ccd45fdc8a162a89af51b87/lib/googleauth/credentials.rb#L357) function can accept a hash. However, that hash will be used to [configure](https://github.com/googleapis/google-auth-library-ruby/blob/2a86821e1b2afc061ccd45fdc8a162a89af51b87/lib/googleauth/credentials.rb#L538) the `Signet::OAuth2::Client` which in turn will [expect an RSA private key](https://github.com/googleapis/google-auth-library-ruby/blob/2a86821e1b2afc061ccd45fdc8a162a89af51b87/lib/googleauth/credentials.rb#L519). Hence, the 💥 error we see above.

## 🙏 Feature request

Let the `Google::Auth::Credentials` initialize function accept a hash with type `external_account`.

## ⚙️ Possible workaround

The only way I spotted to workaround this was to manually build the `Signet::OAuth2::Client` using [`Google::Auth::ExternalAccount::Credentials.make_creds`](https://github.com/googleapis/google-auth-library-ruby/blob/2a86821e1b2afc061ccd45fdc8a162a89af51b87/lib/googleauth/external_account.rb#L40). The downside is that we need to provide a `json_key_io` = we need to transform our credentials hash into a JSON string first (which is not great because `Google::Auth::ExternalAccount::Credentials` [transforms it](https://github.com/googleapis/google-auth-library-ruby/blob/2a86821e1b2afc061ccd45fdc8a162a89af51b87/lib/googleauth/external_account.rb#L56) back to a hash 😿 )

From my example above, here is the workaround:

```ruby
client = ::Google::Cloud::ArtifactRegistry::V1::ArtifactRegistry::Client.new do |config|
  config.credentials = ::Google::Cloud::ArtifactRegistry::V1::ArtifactRegistry::Credentials.new(
    Google::Auth::ExternalAccount::Credentials.make_creds(json_key_io: StringIO.new(JSON.dump(credentials)), scope: 'https://www.googleapis.com/auth/cloud-platform')
  )
end
```

Not great looking code but it works as expected.

**Comment by dazuma** [comment]
Confirmed this issue. The Credentials class appears to be bypassing the normal make_creds mechanism used the the rest of the library to determine which subclass to use. Instead, it tries to reproduce the service account initialization logic, and then uses the signet base class for the oauth client. As a result, it fails to generate external account credentials properly, and also fails to use specialized service account credential functionality such as self-signed JWT credentials.

We'll need to fix Credentials so it uses make_creds instead of trying to reimplement that logic itself. (This should also have the beneficial side effect of removing some duplicated code in credentials.rb.)

**Comment by dazuma** [comment]
This should be fixed in googleauth 1.10.0.
